src/control/tools/EditSelection.cpp

```diff
@@ -690,17 +704,134 @@ void EditSelection::moveSelection(double dx, double dy) {
     this->snappedBounds.x += dx;
     this->snappedBounds.y += dy;
 
-    ensureWithinVisibleArea();
-
     updateMatrix();
 
     this->view->getXournal()->repaintSelection();
 }
 
+void EditSelection::setEdgePan(bool pan) {
+    if (pan && this->edgePanHandler == nullptr) {
+        this->edgePanHandler = g_timeout_source_new(1000 / 30);
+        g_source_set_callback(this->edgePanHandler, reinterpret_cast<int (*)(void*)>(EditSelection::handleEdgePan),
+                              this, nullptr);
+        g_source_attach(this->edgePanHandler, nullptr);
+    } else if (!pan && this->edgePanHandler) {
+        g_source_unref(this->edgePanHandler);
+        this->edgePanHandler = nullptr;
+        this->edgePanInhibitNext = false;
+    }
+}
+
+bool EditSelection::isEdgePanning() const {
+    return this->edgePanHandler;
+}
+
+bool EditSelection::handleEdgePan(EditSelection* self) {
+    Layout* layout = gtk_xournal_get_layout(self->view->getXournal()->getWidget());
+    const double zoom = self->view->getXournal()->getZoom();
+
+    // Layout visible rect and dimensions
+    const int layoutWidth = layout->getMinimalWidth();
+    const int layoutHeight = layout->getMinimalHeight();
+    const auto visRect = layout->getVisibleRect();
+    const double visRectBot = visRect.y + visRect.height;
+    const double visRectRight = visRect.x + visRect.width;
+
+    const auto bbox = self->getBoundingBoxInView();
+    const double bboxBot = bbox.y + bbox.height;
+    const double bboxRight = bbox.x + bbox.width;
+
+    const bool aboveTop = bbox.y < visRect.y;
+    const bool belowBottom = bbox.y + bbox.height > visRectBot;
+    const bool beforeLeft = bbox.x < visRect.x;
+    const bool afterRight = bbox.x + bbox.width > visRect.x + visRect.width;
+
+    // Find scroll direction
+    int scrollYDir = 0;
+    int scrollXDir = 0;
+
+    if (aboveTop) {
+        scrollYDir = -1;
+    } else if (belowBottom) {
+        scrollYDir = 1;
+    }
+
+    if (beforeLeft) {
+        scrollXDir = -1;
+    } else if (afterRight) {
+        scrollXDir = 1;
+    }
+
+    // Compute relative proportion of selection out of view
+    double yScrollMult = 1.0;
+    double xScrollMult = 1.0;
+    const double MAX_MULT = 5.0;
+
+    if (aboveTop) {
+        yScrollMult += (visRect.y - bbox.y) / bbox.height * MAX_MULT;
+    } else if (belowBottom) {
+        yScrollMult += (bboxBot - visRectBot) / bbox.height * MAX_MULT;
+    }
+
+    if (beforeLeft) {
+        xScrollMult += (visRect.x - bbox.x) / bbox.width * MAX_MULT;
+    } else if (afterRight) {
+        xScrollMult += (bboxRight - visRectRight) / bbox.width * MAX_MULT;
+    }
+
+    // How much we'll move the selection and layout.
+    const double SCROLL_SPEED = 5.0;
+    double scrollX = scrollXDir * (SCROLL_SPEED * xScrollMult);
+    double scrollY = scrollYDir * (SCROLL_SPEED * yScrollMult);
+    double layoutScrollX = zoom * scrollX;
+    double layoutScrollY = zoom * scrollY;
```

Comment:
This part of the code confuses me a lot. I don't quite get why `yScrollMult` and `xScrollMult` should be initialized with `1.0`. Should those values stand for the x- and y-relative proportions of the selection out of view? Then these values would be `0.0` under normal circumstances; and if the selection is not moved completely out of view, the compultation would have to yield values between `0.0` and `1.0`.
It should be made 100% clear what `SCROLL_SPEED` and `MAX_MULT` mean. I would interpet these names more according to this suggestion:
```suggestion
    // The maximal speed in pt per update, with which the selection is moved, when it 
    // intersects or at least touches the view.
    const double SCROLL_SPEED = 8.0;
    // The maximal number, with with the scrolling speed can be multiplied, when the 
    // selection is moved far out of view.
    const double MAX_MULT = 3.0;
    
    // Compute relative distance in y- and x-direction of the farthest selection edge 
    // from the view. These values will be between 0.0 and 1.0 when the selection 
    // intersects or touches the view and are bounded by MAX_MULT
    double yScrollMult = 0.0;
    double xScrollMult = 0.0;
 
    if (aboveTop) {
        std::min(yScrollMult = (visRect.y - bbox.y) / bbox.height, MAX_MULT);
    } else if (belowBottom) {
        std::min(yScrollMult = (bboxBot - visRectBot) / bbox.height, MAX_MULT);
    }

    if (beforeLeft) {
        std::min(xScrollMult = (visRect.x - bbox.x) / bbox.width, MAX_MULT);
    } else if (afterRight) {
        std::min(xScrollMult = (bboxRight - visRectRight) / bbox.width, MAX_MULT);
    }

    double scrollX = scrollXDir * (SCROLL_SPEED * xScrollMult);
    double scrollY = scrollYDir * (SCROLL_SPEED * yScrollMult);
    double layoutScrollX = zoom * scrollX;
    double layoutScrollY = zoom * scrollY;```

Reply:
As the panning is currently implemented, the pan amount is calculated as `base amount + bonus amount from out-of-view`. We can definitely change it to just scale linearly instead. I'll have to play around with it to see which feels better first.